crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -77,6 +79,16 @@ pub(crate) struct Checker<'a> {
     deferred: Deferred<'a>,
     // Check-specific state.
     pub(crate) flake8_bugbear_seen: Vec<&'a Expr>,
+    // Dispatchers
+    call_rules: Vec<RegisteredRule<ast::ExprCall>>,
+}
+
+pub(crate) struct RuleContext<'a> {
+    pub(crate) settings: &'a Settings,
+    pub(crate) locator: &'a Locator<'a>,
+    pub(crate) stylist: &'a Stylist<'a>,
+    pub(crate) indexer: &'a Indexer,
+    pub(crate) ctx: &'a Context<'a>,
```

Comment:
Could we add the diagnostics to the `RuleContext`? It may require an additional lifetime but it has the added benefit that the `run` method requires fewer arguments.